### DISCUSSION TITLE
typeerror 'nonetype' object is not iterable

### DISCUSSION THREAD

**Discussion by AlmazKad** [pybricks/support#1945] [Q&A] [answered]
I am building an excavator controlled by an Xbox Controller with two hubs. One in the chassis, the other in the tower.
When i'm trying to transfer data from one hub to another, this error occurs on the receiving hub.

File "Tower.py", line 13, in <module>
TypeError: 'NoneType' object isn't iterable

Have no idea what am i doing wrong....
Here is the chassis code (sending hub, it works completely but i dont know how to check if it is actually sending data):
```
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch
from pybricks.iodevices import XboxController

hub = TechnicHub(broadcast_channel=1)
right = Motor(Port.B)
left = Motor(Port.A)
turntable = Motor(Port.C, Direction.COUNTERCLOCKWISE)
drive_base = DriveBase(left, right, 48, 208)
xbox = XboxController()

while True:
    if xbox.joystick_left()[1] or xbox.joystick_left()[0]:
        drive_base.drive(-xbox.joystick_left()[0] * 2, -xbox.joystick_left()[1] * 2)
    else:
        drive_base.stop()
    if xbox.joystick_right()[0]:
        turntable.run(xbox.joystick_right()[0] * 10)
    else:
        turntable.stop()
    section1 = xbox.joystick_right()[1]
    section2 = 0
    section3 = 0
    if xbox.triggers()[0]:
        section2 = 1
    if xbox.triggers()[1]:
        section3 = 1
    if Button.LB in xbox.buttons.pressed():
        section2 = - 1
    if Button.RB in xbox.buttons.pressed():
        section3 = - 1
    hub.ble.broadcast([section1, section2, section3])
    wait(50)
```
here is the tower code (receiving hub):
```
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch

hub = TechnicHub(observe_channels=[1])
Msection1 = Motor(Port.A)
Msection2 = Motor(Port.B)
Msection3 = Motor(Port.D)

while True:
    section1, section2, section3 = hub.ble.observe(1)   #there is supposedly a mistake in this line
    Msection1.run (section1)
    if section2 == 1:
        Msection2.run (1000)
    if section2 == -1:
        Msection2.run (-1000)
    if section2 == 0:
        Msection2.stop    
    if section3 == 1:
        Msection3.run (1000)
    if section3 == -1:
        Msection3.run (-1000)
    if section3 == 0:
        Msection3.stop
    wait(50)
```

**Comment by dlech** [comment]
> section1, section2, section3 = hub.ble.observe(1)   #there is supposedly a mistake in this line

If the broadcaster has not started yet or has stopped for too long, this will return `None` instead of 3 values. So the correct way to handle it is:

```python
data = hub.ble.observe(1)   

if data is None:
    # handle loss of data
    ...
else:
    section1, section2, section3 = data
    # we have good data
    ...
```

See example: https://docs.pybricks.com/en/stable/hubs/primehub.html#observing-data-from-other-hubs

**Reply by AlmazKad** [reply]
So, i did this, and hub led just keep being red, so no data receiving
```
while True:
    data = hub.ble.observe(1)
    print(data)
    while data is None:
        hub.light.on(Color.RED)
        wait(100)
    else:
        hub.light.on(Color.GREEN)
        section1, section2, section3 = data
```

**Reply by laurensvalk** [reply]
The `while/else` combination is quite unusual. Is that really what you intended?

Your script contains this:

```python
while True:
    data = hub.ble.observe(1)
    print(data)
```

So what did you see on the output?

**Reply by AlmazKad** [reply, marked as answer]
oh, i feel so dumb right now.
well, the code was outputting "none" and going into an infinite loop where new data was not being read. i should have just used the "if" condition, not the "while" loop for checking is there new data. like in the code below.
```
while True:
    data = hub.ble.observe(1)
    print(data)
    if data is None:         #I'm talking about this line
        hub.light.on(Color.RED)
        wait(100)
    else:
        hub.light.on(Color.GREEN)
```
now the data is being sent and received. the led is green. as expected. all that's left is to handle stopping the motors. 
thanks a lot!!

**Reply by laurensvalk** [reply]
Glad it worked :smile:

**Comment by AlmazKad** [comment]
Is it possible that the problem is in the hub hardware? Moreover, other programs are loaded via Bluetooth into them and work properly

**Reply by dlech** [reply]
I suppose it could be. One way you could check if broadcasting is working is to use the [nRF Connect App](https://www.nordicsemi.com/Products/Development-tools/nrf-connect-for-mobile) on a smartphone. The data is just advertising data, so you just have to use the *Scan* feature of the app.

**Reply by AlmazKad** [reply]
I tried the same with two completely different hubs. The situation did not change. I used the app as you recommended, the result is in the screenshots
![photo_2024-11-21_03-02-12](https://github.com/user-attachments/assets/daed048d-ce50-49b9-9b3f-07f9b44f6654)
![photo_2024-11-21_19-51-13](https://github.com/user-attachments/assets/e4051e84-a428-4be0-bec1-35fb3f605820)

**Reply by dlech** [reply]
It looks like these hubs are running official LEGO firmware, not Pybricks firmware?

**Reply by AlmazKad** [reply]
I installed Pybricks firmware on all the hubs I have. Otherwise they wouldn't be controlled by the xbox controller, right?

**Reply by dlech** [reply]
Yes. I guess the app shows the official LEGO stuff for the manufacturer-specific data. So looks like it could be working. 

The `while` loop issue that Laurens pointed out will cause an infinite loop, so probably is the reason why it isn't working.

**Reply by AlmazKad** [reply]
that's right! thank you too for your help!
